keps/sig-scheduling/5278-nominated-node-name-for-expectation/README.md

```diff
@@ -0,0 +1,983 @@
+
+## Proposal
+
+### User Stories (Optional)
+
+Here is the all use cases of NominatedNodeNames that we're taking into consideration:
+- The scheduler puts it after the preemption (already implemented)
+- The scheduler puts it at the beginning of binding cycles (only if the binding cycles invole PreBind phase)
+- The cluster autoscaler puts it after creating a new node for pending pod(s) so that the scheduler can find a place faster when the node is created.
+- Kueue uses it to determine a prefered node for the pod based on their internal calculation (Topology aware scheduling etc)
+
+(Possibly, our future initiative around the workload scheduling (including gang scheduling) can also utilize it,
+but we don't discuss it here because it's not yet concreted at all.)
+
+#### Story 1: Prevent inappropriate scale downs by Cluster Autoscaler 
+
+The scheduler starts to expose where the pod is going to with `NominatedNodeName` at the beginning of binding cycles.
+And, the cluster autoscaler takes `NominatedNodeName` into consideration when calculating which nodes they delete.
+
+It helps the scenarios where the binding cycles take time, for example, VolumeBinding plugin takes time at PreBind extension point.
+
+#### Story 2: Cluster Autoscaler specifies `NominatedNodeName` to indicate where pods can go after new nodes are created/registered
+
+Usually, the scheduler scans all the nodes in the cluster when scheduling pods.
+
+When the cluster autoscaler creates instances for pending pods, it calculate which new node might get which pending pod.
+If they can put `NominatedNodeName` based on those calculation, it could tell the scheduler that the node can probably picked up for the pod's scheduling,
+prevenging the double effort of scanning/calculating all nodes again at the scheduling retries.
+
+#### Story 3: Kueue specifies `NominatedNodeName` to indicate where it prefers pods being scheduled to
+
+When Kueue determines where pods are prefered to being scheduled on, based on their internal scheduling soft constraints (Preferred Topology Aware Scheduling, etc)
+currently, they just put the node selector to tell the scheduler about their preference, and then un-gate the pods.
+
+After this proposal, they can specify `NominatedNodeName` instead of a prefered node selector, 
+which makes the probability of pods being scheduled onto the node higher.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Increasing the load to kube-apiserver
+
+If we simply implement this, we'd double the API calls during a simple binding cycle (NNN + actual binding),
```

Comment:
Should we be concerned about additional load on apiserver coming from Cluster Autoscaler? Most of the time CA will generate much lower load than scheduler, but in case of failing a large cluster scale up there will be a spike of NNN setting, followed by NNN clearing.

Reply:
@kubernetes/sig-scalability @wojtek-t 
Any opinion on this? In short, the cluster autoscaler will want to put NNN on pods after scaling up nodes for them, and if the node creation somehow fails, it has to clear NNN or update NNN with a new value. So, those would be additional API calls from the cluster autoscaler.